src/redirects.rs

```diff
@@ -6,7 +6,106 @@
 //! Redirection module to handle config redirect URLs with pattern matching support.
 //!
 
-use crate::settings::Redirects;
+use headers::HeaderValue;
+use hyper::{Body, Request, Response, StatusCode};
+
+use crate::{error_page, handler::RequestHandlerOpts, settings::Redirects, Error};
+
+pub(crate) fn pre_process(
+    opts: &RequestHandlerOpts,
+    req: &Request<Body>,
+) -> Option<Result<Response<Body>, Error>> {
+    let redirects = opts.advanced_opts.as_ref()?.redirects.as_deref()?;
+
+    let uri = req.uri();
+    let uri_path = uri.path();
+    let host = req
+        .headers()
+        .get(http::header::HOST)
+        .and_then(|v| v.to_str().ok())
+        .unwrap_or("");
+    let mut uri_host = uri.host().unwrap_or(host).to_owned();
+    if let Some(uri_port) = uri.port_u16() {
+        uri_host.push_str(&format!(":{}", uri_port));
+    }
+    if let Some(matched) = get_redirection(&uri_host, uri_path, Some(redirects)) {
+        // Redirects: Handle replacements (placeholders)
+        let regex_caps = if let Some(regex_caps) = matched.source.captures(uri_path) {
+            regex_caps
+        } else {
+            return handle_error(
+                "unexpected regex failure",
+                "extracting captures failed",
+                opts,
+                req,
+            );
+        };
+
+        let caps_range = 0..regex_caps.len();
+        let caps = caps_range
+            .clone()
+            .map(|i| regex_caps.get(i).map(|s| s.as_str()).unwrap_or(""))
+            .collect::<Vec<&str>>();
+
+        let patterns = caps_range
+            .map(|i| format!("${}", i))
+            .collect::<Vec<String>>();
+
+        let dest = &matched.destination;
+
+        tracing::debug!("url redirects glob pattern: {:?}", patterns);
+        tracing::debug!("url redirects regex equivalent: {}", matched.source);
+        tracing::debug!("url redirects glob pattern captures: {:?}", caps);
+        tracing::debug!("url redirects glob pattern destination: {:?}", dest);
+
+        let ac = match aho_corasick::AhoCorasick::new(patterns) {
+            Ok(ac) => ac,
+            Err(err) => {
+                return handle_error("failed creating Aho-Corasick matcher", err, opts, req)
+            }
+        };
+        let dest = match ac.try_replace_all(dest, &caps) {
+            Ok(dest) => dest.to_string(),
+            Err(err) => return handle_error("failed replacing captures", err, opts, req),
+        };
+
+        tracing::debug!(
+            "url redirects glob pattern destination replaced: {:?}",
+            dest
+        );
```

Comment:
Just a comment. The placeholder replacement logic over here is almost identical to the one for URL rewrites. So we could think of moving it to a commonplace.

Reply:
Ok. I’ll be moving the rewrite logic as the next thing anyway, and I’ll see whether I can merge some of the code.